File: src/0058_length_of_last_word.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLastWord = function (s) {
    let len = 0;

    for (let i in s) {
        if (s[i] != ' ') {
            if (s[i - 1] == ' ') len = 1;
            else len += 1;
        }
    }
    return len;
};

// another approach. starting out from the last so we don't have to go all the way to the end.
var lengthOfLastWord = function (s) {
    let firstCharOccurance = false;
    let lastWordLen = 0;

    for (let i = s.length - 1; i > -1; i--) {
        if (s[i] !== ' ') {
            firstCharOccurance = true;
            lastWordLen++;
        }
        if (firstCharOccurance && s[i] === ' ') {
            break;
        }
    }
    return lastWordLen;
};

module.exports = { lengthOfLastWord };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(lengthOfLastWord("Hello World"), 5);
    assert.deepStrictEqual(lengthOfLastWord("   fly me   to   the moon  "), 4);
    assert.deepStrictEqual(lengthOfLastWord("luffy is still joyboy"), 6);
    assert.deepStrictEqual(lengthOfLastWord("a"), 1);
}

File: src/0058_length_of_last_word.rs
struct Solution;

impl Solution {
    pub fn length_of_last_word(s: String) -> i32 {
        let s = s.trim_end();
        let s: Vec<char> = s.chars().collect();
        let mut ans = 0;
        for i in (0..s.len()).rev() {
            if !s[i].is_whitespace() {
                ans += 1;
            } else {
                break;
            }
        }
        ans
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            Solution::length_of_last_word(String::from("Hello World")),
            5
        );
        assert_eq!(
            Solution::length_of_last_word(String::from("   fly me   to   the moon  ")),
            4
        );
        assert_eq!(
            Solution::length_of_last_word(String::from("luffy is still joyboy")),
            6
        );
        assert_eq!(Solution::length_of_last_word(String::from("a")), 1);
    }
}
